ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -100,20 +111,50 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
    * @throws CloudStorageException if the upload failed.
    */
   private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
-      throws CloudStorageException, IOException {
+      throws CloudStorageException, IOException, GeneralSecurityException {
     if (shouldUpload(messageInfo)) {
       BlobId blobId = (BlobId) messageInfo.getStoreKey();
-      if (cloudConfig.vcrRequireEncryption && !BlobId.isEncrypted(blobId.getID())) {
-        // TODO: encrypt on demand before uploading, skip for now
-        return;
+      boolean isRouterEncrypted = isRouterEncrypted(blobId);
+      String kmsContext = null;
+      EncryptionOrigin encryptionOrigin = isRouterEncrypted ? EncryptionOrigin.ROUTER : EncryptionOrigin.NONE;
+      if (cloudConfig.vcrRequireEncryption) {
+        if (isRouterEncrypted) {
+          // Nothing further needed
+        } else {
+          // Need to encrypt the buffer before upload
+          if (cryptoAgent != null) {
+            Timer.Context encryptionTimer = vcrMetrics.blobEncryptionTime.time();
+            messageBuf = cryptoAgent.encrypt(messageBuf);
+            encryptionTimer.stop();
+            vcrMetrics.blobEncryptionCount.inc();
+            encryptionOrigin = EncryptionOrigin.VCR;
+            kmsContext = cryptoAgent.getEncryptionContext();
+          } else {
+            // Cannot upload this blob since we can't satisfy encryption requirements
+            vcrMetrics.skipUnencryptedBlobsCount.inc();
+            return;
+          }
+        }
+      } else {
+        // encryption not required, upload as is
       }
       CloudBlobMetadata blobMetadata =
           new CloudBlobMetadata(blobId, messageInfo.getOperationTimeMs(), messageInfo.getExpirationTimeInMs(),
-              messageInfo.getSize());
+              messageInfo.getSize(), encryptionOrigin, kmsContext);
```

Comment:
Is there a reason why not add encryption algorithm or cryptoAgent? 

If we are required to change encryption algorithm, there should be a way to know blobs already uploaded were with different algorithm. So I think either encryption algorithm or cryptoAgent should be added as metadata for backwards compatibility. It can be stored in payload or DB. 

I prefer to store agentFactory in payload, so that when a blob is being get, `CloudBlobStore` can create `deCryptoAgent` while parsing payload.

Reply:
The problem is, we don't necessarily even know the encryption algorithm being used.  In some cases the implementation details are hidden.
Storing the agentFactory name in the payload is problematic since it's a variable length string (another size we need to add to header).  We could add it to metadata, but I don't see a great advantage.

Reply:
I added cryptoAgentFactory to metadata.

Reply:
cool.